2.0/m_joinpartspam.cpp

```diff
@@ -109,114 +157,225 @@ class joinpartspamsettings
 			else
 				++it;
 		}
+
+		for (std::map<std::string, time_t>::iterator i = blocked.begin(); i != blocked.end(); )
+		{
+			if (ServerInstance->Time() > i->second)
+				blocked.erase(i++);
+			else
+				++i;
+		}
 	}
 };
 
-/* Handles channel mode +V */
 class JoinPartSpam : public ModeHandler
 {
+	SimpleExtItem<joinpartspamsettings>& ext;
+	bool& allowredirect;
+	bool& freeredirect;
+
+	bool ParseCycles(irc::sepstream& stream, unsigned int& cycles)
+	{
+		std::string strcycles;
+		if (!stream.GetToken(strcycles))
+			return false;
+
+		unsigned int result = ConvToInt(strcycles);
+		if (result < 2 || result > 20)
+			return false;
+
+		cycles = result;
+		return true;
+	}
+
+	bool ParseSeconds(irc::sepstream& stream, unsigned int& seconds)
+	{
+		std::string strseconds;
+		if (!stream.GetToken(strseconds))
+			return false;
+
+		unsigned int result = ConvToInt(strseconds);
+		if (result < 1 || result > 43200)
+			return false;
+
+		seconds = result;
+		return true;
+	}
+
+	bool ParseRedirect(irc::sepstream& stream, std::string& redirect, User* source, Channel* chan)
+	{
+		std::string strredirect;
+		// This parameter is optional
+		if (!stream.GetToken(strredirect))
+			return true;
+
+		if (!allowredirect)
+		{
+			source->WriteNumeric(ERR_INVALIDPARM, "%s %s %s :Invalid join/part spam parameter, the server admin has disabled channel redirection.",
```

Comment:
Can you use ERR_INVALIDMODEPARAM from master instead? The syntax is:

     :<source> 696 <targetnick> <channel> <modechar> <parameter> :Invalid <name> mode parameter.